opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py

```diff
@@ -349,6 +362,316 @@ def collect(
+        # https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#exponential-bucket-histogram-aggregation)
+        max_size: int = 160,
+    ):
+        super().__init__(attributes)
+        # max_size is the maximum capacity of the positive and negative
+        # buckets.
+        if max_size < self._min_max_size:
+            raise Exception("size {max_size} is smaller than {min_max_size}")
+
+        if max_size > self._max_max_size:
+            raise Exception("size {max_size} is larter than {max_max_size}")
+
+        self._max_size = max_size
+
+        # _sum is the sum of all the values aggregated by this aggregator.
+        self._sum = 0
+
+        # _count is the count of all calls to aggregate.
+        self._count = 0
+
+        # _zero_count is the count of all the calls to aggregate when the value
+        # to be aggregated is exactly 0.
+        self._zero_count = 0
+
+        # _min is the smallest value aggregated by this aggregator.
+        self._min = inf
+
+        # _max is the smallest value aggregated by this aggregator.
+        self._max = -inf
+
+        # _positive holds the positive values.
+        self._positive = Buckets()
+
+        # _negative holds the negative values by their absolute value.
+        self._negative = Buckets()
+
+        # _mapping corresponds to the current scale, is shared by both the
+        # positive and negative buckets.
+        self._mapping = LogarithmMapping(LogarithmMapping._max_scale)
+
+        self._instrument_temporality = AggregationTemporality.DELTA
+        self._start_time_unix_nano = start_time_unix_nano
+
+    def aggregate(self, measurement: Measurement) -> None:
+        # pylint: disable=too-many-branches,too-many-statements
+
+        value = measurement.value
+
+        if value < self._min:
+            self._min = value
+
+        if value > self._max:
+            self._max = value
+
+        self._count += 1
+
+        if value == 0:
+            self._zero_count += 1
+            return
+
+        self._sum += value
+
+        if value > 0:
+            buckets = self._positive
+
+        else:
+            value = -value
+            buckets = self._negative
+
+        index = self._mapping.map_to_index(value)
+
+        is_rescaling_needed = False
+
+        if buckets.len() == 0:
+            buckets._index_start = index
+            buckets._index_end = index
+            buckets._index_base = index
+
+        elif (
+            index < buckets._index_start
+            and (buckets._index_end - index) >= self._max_size
+        ):
+            is_rescaling_needed = True
+            low = index
+            high = buckets._index_end
+
+        elif (
+            index > buckets._index_end
+            and (index - buckets._index_start) >= self._max_size
+        ):
+            is_rescaling_needed = True
+            low = buckets._index_start
+            high = index
+
+        if is_rescaling_needed:
```

Comment:
Suggest breaking out the logic of this branch into a separate method, it's getting long...

Reply:
I first had it split as it is [here](https://github.com/lightstep/otel-launcher-go/blob/c9ca8483be067a39ab306b09060446e7fda65f35/lightstep/sdk/metric/aggregator/histogram/structure/exponential.go#L361) but it was hard reading it since `incrementIndexBy` actually may increment the index or not.